### PR TITLE
Warn early that overwrite-policy is needed before build

### PR DESCRIPTION
**Pull request by philderbeast**
See #5993.

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary (not necessary).
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Bonus points for added automated tests!

### CODE REVIEW

**Comment by philderbeast** [comment]
To test, unpack a package from hackage that you've previously installed and try installing it without an `overwrite-policy`. It should stop with a warning before building.

```
$ cabal get updo-1.0.0
Unpacking to updo-1.0.0/

$ cd updo-1.0.0/

$ cabal --version
cabal-install version 3.11
compiled using version 3.11.0.0 of the Cabal library 

$ cabal install
Wrote tarball sdist to
/.../updo-1.0.0/dist-newstyle/sdist/updo-1.0.0.tar.gz
Resolving dependencies...
Build profile: -w ghc-9.4.6 -O1
In order, the following will be built (use -v for more details):
 - updo-1.0.0 (exe:updo-pkg-groups) (requires build)
 - updo-1.0.0 (exe:updo-pkgs-sorted) (requires build)
 - updo-1.0.0 (exe:updo-pkgs-upgrade-done) (requires build)
 - updo-1.0.0 (exe:updo-pkgs-upgrade-partition) (requires build)
 - updo-1.0.0 (exe:updo-sha256map) (requires build)
Error: cabal: Path '/.../.cabal/bin/updo-pkg-groups' already
exists. Use --overwrite-policy=always to overwrite.

$ cabal install --overwrite-policy=always
Wrote tarball sdist to
/.../updo-1.0.0/dist-newstyle/sdist/updo-1.0.0.tar.gz
Resolving dependencies...
Build profile: -w ghc-9.4.6 -O1
In order, the following will be built (use -v for more details):
 - updo-1.0.0 (exe:updo-pkg-groups) (requires build)
 - updo-1.0.0 (exe:updo-pkgs-sorted) (requires build)
 - updo-1.0.0 (exe:updo-pkgs-upgrade-done) (requires build)
 - updo-1.0.0 (exe:updo-pkgs-upgrade-partition) (requires build)
 - updo-1.0.0 (exe:updo-sha256map) (requires build)
Starting     updo-1.0.0 (exe:updo-pkg-groups)
Starting     updo-1.0.0 (exe:updo-pkgs-upgrade-done)
Starting     updo-1.0.0 (exe:updo-pkgs-upgrade-partition)
Starting     updo-1.0.0 (exe:updo-pkgs-sorted)
Starting     updo-1.0.0 (exe:updo-sha256map)
Building     updo-1.0.0 (exe:updo-pkgs-upgrade-done)
Building     updo-1.0.0 (exe:updo-pkgs-upgrade-partition)
Building     updo-1.0.0 (exe:updo-pkg-groups)
Building     updo-1.0.0 (exe:updo-pkgs-sorted)
Building     updo-1.0.0 (exe:updo-sha256map)
Installing   updo-1.0.0 (exe:updo-pkgs-upgrade-done)
Installing   updo-1.0.0 (exe:updo-pkgs-sorted)
Completed    updo-1.0.0 (exe:updo-pkgs-upgrade-done)
Completed    updo-1.0.0 (exe:updo-pkgs-sorted)
Installing   updo-1.0.0 (exe:updo-pkgs-upgrade-partition)
Completed    updo-1.0.0 (exe:updo-pkgs-upgrade-partition)
Installing   updo-1.0.0 (exe:updo-pkg-groups)
Completed    updo-1.0.0 (exe:updo-pkg-groups)
Installing   updo-1.0.0 (exe:updo-sha256map)
Completed    updo-1.0.0 (exe:updo-sha256map)
Symlinking 'updo-pkg-groups' to
'/.../.cabal/bin/updo-pkg-groups'
Symlinking 'updo-pkgs-sorted' to
'/.../.cabal/bin/updo-pkgs-sorted'
Symlinking 'updo-pkgs-upgrade-done' to
'/.../.cabal/bin/updo-pkgs-upgrade-done'
Symlinking 'updo-pkgs-upgrade-partition' to
'/.../.cabal/bin/updo-pkgs-upgrade-partition'
Symlinking 'updo-sha256map' to '/.../.cabal/bin/updo-sha256map'
```

**Review by geekosaur — APPROVED** [review]

**Comment by philderbeast** [comment]
@Mikolaj the check failures don't seem related to changes I've made. Is there anything I need to do to move these along?

**Comment by Mikolaj** [comment]
I "fixed" precisely that one yesterday in another PR by clearing a part of the CI cache. Let me restart your CI and if it doesn't help, I'd clear some other parts of the cache.

**Comment by ulysses4ever** [comment]
Great contribution! I'll try to review properly as soon as I'm free from some pressing stuff. Hope someone beats me to it though.

**Review by andreabedini — DISMISSED** [review]
At a first glance this looks very good! I'll review it in more detail tomorrow; in the meantime it needs tests ☺️

**Comment by andreabedini** [comment]
I see tests and dismiss my request for changes. I didn't have the bandwidth to look at this unfortuntately.

File: changelog.d/issue-5993
Comment (ulysses4ever) [inline changelog.d/issue-5993:1]: Currently it sounds a bit cryptic and rough. This could be expanded to explain the issue a bit better. Either right in `sinopsis` and/or in `description`.

**Review by geekosaur — APPROVED** [review]

File: cabal-testsuite/PackageTests/LinkerOptions/NonignoredConfigs/cabal.test.hs
Comment (geekosaur) [inline cabal-testsuite/PackageTests/LinkerOptions/NonignoredConfigs/cabal.test.hs:74]: I think you need to tell it to reinstall here to avoid the validate error?

**Comment by philderbeast** [comment]
Could I please be allowed to add labels? I would like to mark this as a WIP. I'm finding that adding the test is harder than making the fix and don't want to bother reviewers with something half done.

**Comment by ffaf1** [comment]
@philderbeast there is a “Still in progress? convert to draft” button on the right bar of the UI. Would that do?

File: cabal-testsuite/PackageTests/LinkerOptions/NonignoredConfigs/cabal.test.hs
Comment (philderbeast) [reply cabal-testsuite/PackageTests/LinkerOptions/NonignoredConfigs/cabal.test.hs]: Yes sorry @geekosaur, I pushed a work in progress experiment to see if it would fail before adding the flag for reinstalling.

File: changelog.d/issue-5993
Comment (philderbeast) [reply changelog.d/issue-5993:1]: Thanks for suggesting this. I added a description.

Comment (ulysses4ever) [reply changelog.d/issue-5993:1]: Thanks!

**Comment by philderbeast** [comment]
It looks like the normalization of the test output doesn't give the same result when running on Windows for CI run `Validate / Validate windows-latest ghc-9.4.4 (pull_request)`:

```diff
Actual output differs from expected:

stderr:
--- "D:\\a\\cabal\\cabal\\cabal-testsuite\\PackageTests\\WarnEarlyOverwrite\\clean-install-by-copy.dist\\clean-install-by-copy.out.normalized"	2023-10-30 22:50:07.370377400 +0000
+++ "D:\\a\\cabal\\cabal\\cabal-testsuite\\PackageTests\\WarnEarlyOverwrite\\clean-install-by-copy.dist\\clean-install-by-copy.comp.out.normalized"	2023-10-30 22:50:07.370377400 +0000
@@ -8,5 +8,5 @@
 Preprocessing executable 'warn-early-overwrite' for WarnEarlyOverwrite-0.1.0.0...
 Building executable 'warn-early-overwrite' for WarnEarlyOverwrite-0.1.0.0...
 Installing executable warn-early-overwrite in <PATH>
-Warning: The directory <GBLTMPDIR>/ghc-<GHCVER>/incoming/new-<RAND><GBLTMPDIR>/ghc-<GHCVER>/<PACKAGE>-<HASH>/bin is not in the system search path.
-Copying 'warn-early-overwrite' to '<GBLTMPDIR>/warn-early-overwrite'
+Warning: The directory <GBLTMPDIR><GHCVER>/incoming/new-2448/Users/RUNNER~1/AppData/Local/Temp/cabal-test-store-28260/ghc-<GHCVER>/WarnEarlyOver_-0.1.0.0-4c19059e06a32b93b2812983631117e77a2d3833/bin is not in the system search path.
+Copying 'warn-early-overwrite' to '<GBLTMPDIR>'
*** Exception: ExitFailure 1
```

**Comment by philderbeast** [comment]
I've added tests for this feature.

**Comment by ulysses4ever** [comment]
I’m currently a little over my time budget for Cabal stuff but I’ll try to get to it this week, hopefully. (I’m happy if someone beats me to though)

**Comment by Mikolaj** [comment]
Is there a way to test this on Windows? Via manual QA tests perhaps?

**Comment by philderbeast** [comment]
> Is there a way to test this on Windows?

The other option is to open another follow on issue where we try to normalize the output so we don't have to `skipIfWindows` for these tests.

**Comment by Mikolaj** [comment]
> > Is there a way to test this on Windows?
> 
> The other option is to open another follow on issue where we try to normalize the output so we don't have to `skipIfWindows` for these tests.

Oh, there is more such tests than the one you added (note that at least some Windows tests are disabled due to other reasons, e.g., transient linker segfaults with newest GHCs)? In that case your proposal sounds good to me.

**Review by ulysses4ever — APPROVED** [review]
This is a bit above my paygrade, but it looks like a high-quality patch superficially. And it kills an ages-standing wart in the UX. So, I'm happy to +1 it. I'm sorry it took me so long, nut I really hoped that someone more qualified could give it a good review. I will say that I'd prefer to see haddocks on all new functions, even more trivial ones.

@mpickering maybe you could give this one a quick look?

File: cabal-install/src/Distribution/Client/CmdInstall.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/CmdInstall.hs:1107]: It's a little sad to see a haddock just sayong "check X" on a function returning `IO ()`. I'd rather have separate pure checking function or a longer haddock explaining what happens if check succeeds/fails.

File: cabal-install/src/Distribution/Client/CmdInstall.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/CmdInstall.hs:862]: Same as with `installableUnitExes`. Also, typos (were there before):
```suggestion
-- | Can we install and build executable by symlinking or copying it?
```

Comment (geekosaur) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:862]: The original actually looks correct to me, aside from the abbreviation. Read "any built exe" as "any executable that was built".

Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:862]: That's fair! Could we rephrase it like you say? I can understand what you say but I completely missed what the original was saying...

Comment (geekosaur) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:862]: Sure.
```suggestion
-- | Can we install any executable that was built by symlinking or copying it?
```

Comment (philderbeast) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:862]: Thanks for the suggestions. I didn't change "exe" to "executable" but I did change "symlinking/copying" to "symlinking or copying" everywhere, including in one user-facing message. I 'lso changed "symlink/copy" to "symlink or copy".

File: cabal-install/src/Distribution/Client/CmdInstall.hs
Comment (philderbeast) [reply cabal-install/src/Distribution/Client/CmdInstall.hs]: I improved the message from "check" to "try" and "stop".

**Comment by philderbeast** [comment]
> I'd prefer less commits

Yes @ulysses4ever, I can squash some of the commits.

**Comment by philderbeast** [comment]
> I can have a look tomorrow. I find it a little difficult to keep track of what are refactorings and what is needed for functionality changes from a quick glance though.

Thanks @mpickering. I've squashed commits and rejigged the changes to minimize differences. @ulysses4ever you may want to take another look.

**Comment by ulysses4ever** [comment]
Is there no way to have two separate commits, one for refactoring and one for the behavioral change? It's fine if not, just asking.

**Comment by geekosaur** [comment]
It's possible but a pain, especially if the code changes are interleaved (which means you can't use git's ability to pick individual diff chunks).

**Comment by philderbeast** [comment]
@Mikolaj should I add the `merge+no-rebase` label now?

**Comment by mpickering** [comment]
I have looked at this for a little bit but the diff is just too hard for me to read and understand without putting more effort in. I found it hard to work out what the functional changes were, what the refactoring was or what was an artifact introduced by formatting so I won't be able to review this one.

**Comment by Mikolaj** [comment]
@philderbeast: yes please, as soon as you are satisfied with the process. Beyond 2 reviews it's your call whether to ask for more.

**Comment by philderbeast** [comment]
> I have looked at this for a little bit but the diff is just too hard for me to read and understand

@mpickering I read that as negative review.

Overall, I found adding tests for this fix was a lot more time and work than doing the fix. The docs on [how to write tests](https://github.com/haskell/cabal/tree/master/cabal-testsuite#how-to-write) helped.

When I first put this up for review, I'd moved some functions to top level by need so that I could use them in two places. I had not placed these strategically to reduce the diff count. On request, I squashed commits but also moved lines back closer to where they'd been to minimize diffs (some diffs I couldn't minimize because the formatter would change the layout again, mostly indentation changes). Sorry I didn't think to minimize diffs before asking for review.

What could count as a refactoring is collecting repeated series of function arguments into records, `InstallCfg` and `InstallExe`. I feel the declutter there is worth it.

**Comment by philderbeast** [comment]
To run the test locally:

```
$ ghcup set ghc 9.4.7
$ cabal build all --enable-tests --enable-benchmarks
$ cabal run cabal-testsuite:cabal-tests -- --with-cabal=./dist-newstyle/build/x86_64-linux/ghc-9.4.7/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal cabal-testsuite/PackageTests/WarnEarlyOverwrite/dirty-install.test.hs
...
-----BEGIN CABAL OUTPUT-----
Installing executable warn-early-overwrite in /tmp/cabal-test-store-62007/ghc-9.4.7/incoming/new-62101/tmp/cabal-test-store-62007/ghc-9.4.7/WarnEarlyOverwrite-0.1.0.0-e-warn-early-overwrite-81f111c4ea95736fd5a3f16b46b053fdec9c718a890fbe97dc2b5338f6a7de04/bin
-----END CABAL OUTPUT-----
-----BEGIN CABAL OUTPUT-----
Warning: The directory /tmp/cabal-test-store-62007/ghc-9.4.7/incoming/new-62101/tmp/cabal-test-store-62007/ghc-9.4.7/WarnEarlyOverwrite-0.1.0.0-e-warn-early-overwrite-81f111c4ea95736fd5a3f16b46b053fdec9c718a890fbe97dc2b5338f6a7de04/bin is not in the system search path.
-----END CABAL OUTPUT-----
...
-----BEGIN CABAL OUTPUT-----
Symlinking 'warn-early-overwrite' to '/tmp/cabal-test-store-62007/warn-early-overwrite'
-----END CABAL OUTPUT-----
OK

**Comment by ulysses4ever** [comment]
> What could count as a refactoring is collecting repeated series of function arguments into records, InstallCfg and InstallExe. I feel the declutter there is worth it.

definitely refractoring.

> moved some functions to top level by need 

Same.

Would be helpful for potential reviewers, I think, to have these in a separate commit. But that's okay.

Thank you for the contribution!

**Comment by philderbeast** [comment]
Closed in favour of #9506.

File: cabal-install/src/Distribution/Client/CmdInstall.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/CmdInstall.hs:1000]: Haddock?

File: cabal-install/src/Distribution/Client/InstallSymlink.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/InstallSymlink.hs:157]: Stuff like this may be shortened by RecordWildcards but some may say that it's less readable, so up to you.

File: cabal-install/src/Distribution/Client/InstallSymlink.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/InstallSymlink.hs:268]: A little too terse to my taste.

File: cabal-install/src/Distribution/Client/CmdInstall.hs
Comment (philderbeast) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:1000]: I've added haddocks.

File: cabal-install/src/Distribution/Client/InstallSymlink.hs
Comment (philderbeast) [reply cabal-install/src/Distribution/Client/InstallSymlink.hs:268]:
```diff
- -- | How to handle symlinking a binary.
+ -- | After checking if a target is writeable given the overwrite policy,
+ -- dispatch to an appropriate action;
+ --  * @onMissing@ if the target doesn't exist
+ --  * @onOverwrite@ if the target exists and we are allowed to overwrite it
+ --  * @onNever@ if the target exists and we are never allowed to overwrite it
+ --  * @onPrompt@ if the target exists and we are allowed to overwrite after prompting
```

File: cabal-install/src/Distribution/Client/InstallSymlink.hs
Comment (philderbeast) [reply cabal-install/src/Distribution/Client/InstallSymlink.hs:157]: I normally shy away from `-XRecordWildCards` but in this case it looks fine so I went with it. Thanks for suggesting it @ulysses4ever.

Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/InstallSymlink.hs:157]: Right, I'm also cautious about RWC but this looked like a good spot for it. Thank you for acting on the feedback!

**Pull request closed without merging**
